### DISCUSSION TITLE
Nullable planning values in chained models?

### DISCUSSION THREAD

**Discussion by holgerbrandl** [TimefoldAI/timefold-solver#754] [General]
Hi there,

from the documentation we have learned that in chained models planning values must be set and can not be nullable. Is there a way/workaround to bypass this restriction? 
There are use cases where some items can not be planned (e.g. because of lacking tool capabilities in production scheduling). We could prefilter these planning entities, but we also encounter use cases where an unassigned production task is better planned (in terms of overall score) without a tool assignment. So only during the solving process, it is determined that the planning value should not be set to optimize the solution.

Kind regards,
Holger

**Comment by triceo** [comment]
Hello @holgerbrandl!

The solution is called [virtual values](https://docs.timefold.ai/timefold-solver/latest/responding-to-change/responding-to-change#overconstrainedPlanningWithVirtualValues). Essentially, where you'd use `null`, use a dummy entity. This has a lot of consequences for your model - constraints need to skip the dummy entity, some moves will have to set up filtering etc.

The recently released Timefold Solver 1.8 solves this problem [for list variables](https://docs.timefold.ai/timefold-solver/latest/using-timefold-solver/modeling-planning-problems#planningListVariableAllowingUnassigned), and I'd like to encourage you to switch - that model is faster, easier to understand and (unlike chaining) it will be further enriched with features in the future.

**Reply by holgerbrandl** [reply]
Thanks @triceo .  Indeed, we have considered virtual values and stumbled over the complexities mentioned in the documentation. We have also worked with list variable models, and love their simplicity. It's good to know that `null` is supported there already. 

To prepare a migration: What features are yet missing in list variables, is there a listing in the tracker or the docs, or are both (chained and list model) en par regarding supported features?

**Reply by triceo** [reply]
We keep a list in the [list variable section](https://docs.timefold.ai/timefold-solver/latest/using-timefold-solver/modeling-planning-problems#planningListVariable) of the documentation.
